cargodocs/centralized.md

```diff
@@ -0,0 +1,110 @@
+Telescope construction with prehashing and bounded DFS scheme given in _Section 3.2.2_.
+
+
+In some cases, $n_p$ may be small, which can affect the completeness and efficiency of the proof construction. The protocol adapts to such cases by introducing optimizations and tighter bounds.
+- In the basic ALBA scheme, optimal performance requires $n_p \geq \lambda^3$.
+- When $n_p$ is small (e.g., $n_p < \lambda^3$), the probability of successfully finding valid proof paths diminishes.
+- Ensure that the proof remains efficient and sound even when $n_p$ is small by introducing an improved protocol with tighter guarantees on the number of attempts and running time.
+- When $n_p$ is too small to guarantee a valid proof in a single attempt, the protocol allows the prover to make multiple attempts.
+- Each attempt is indexed by an integer $v \in \{1, \ldots, r\}$, where $r$ is the number of repetitions. This ensures higher completeness by amplifying the success probability.
+  - Set $r := \lambda_{\text{rel}}$, ensuring the completeness error becomes $2^{-\lambda_{\text{rel}}}$.
+  - This approach maintains completeness even when the prover has a limited number of elements in their set.
+
+### Optimized Proof Construction
+- A valid proof now consists of a tuple $(v, t, s_1, \ldots, s_u)$, where:
+  - $v$ is the repetition index.
+  - $t \in \{1, \ldots, d\}$ is another index used to explore proof paths.
+  - $s_1, \ldots, s_u$ are elements from the prover's set.
+- The tuple must satisfy:
+  $$
+  H_1(v, t, s_1, \ldots, s_{i-1}) = H_0(v, s_i), \quad \forall i \in \{1, \ldots, u\}
+  $$
+  $$
+  H_2(v, t, s_1, \ldots, s_u) = 1
+  $$
+
+### Performance
+- **Proof Size:** $u = \left\lceil \frac{\lambda_{\text{sec}} + \log \lambda_{\text{rel}} + 5 - \log \log e}{\log \left(\frac{n_p}{n_f}\right)} \right\rceil$
+- **Number of Repetitions:**  $r := \lambda_{\text{rel}}$ ensures completeness with error $2^{-\lambda_{\text{rel}}}$.
+- **Expected Running Time:** The average running time for the prover is: $n_p + \mathcal{O}(u^2)$.
+- **Worst-Case Running Time:** The worst-case running time is bounded by: $n_p + \mathcal{O}(u^2 \cdot \lambda_{\text{rel}})$.
+
+## Algorithm: Main Construction
+
+### Parameters
+- Soundness security parameter: $\lambda_{sec}$
+- Completeness security parameter: $\lambda_{rel}$
+- Approximate size of set $S_p$ to lower bound: $n_p$
```

Comment:
@tolikzinovyev : I suppose this is the parameter that causes most of the confusion for anyone who tries to apply Alba. I do understand the motivation from a complexity theory perspective (the original Babai's work - language recognition, approx upper/lower bound problem cannot be stated as a language recognition problem, etc.) but is going that deep in the complexity theory helping the user of Alba (plus, these concepts of Game of Nature, etc. are highly non-obvious to a non-expert on complexity theory)? For the purpose of the parameters, I'd advocate, as much as possible, to simplify the setting.

Reply:
In short, what exactly do I gain by having this parameter `n_p` as opposed to just working with `S_p`?

Reply:
As I said in another thread, in the paper we simply have n_p = |S_p|. We can consider |S_p| >= n_p since having more elements might in practice speed up the prover, but we need to be careful not to have |S_p| much larger than n_p because that would actually slow down the prover (the balls to bin hashing becomes the bottleneck). Also in the paper we assumed |S_p| = n_p when analyzing the prover running time. The DFS seems to terminate earlier in practice when |S_p| > n_p and it seems intuitive but we don't have a proof that it does not terminate later. I don't know how strict we want to be here, it's probably ok to just require n_p <= |S_p| <= 2 * n_p.